# Decision Tree | Assignment

Assignment Code: DA-AG-012



# Question 1: What is a Decision Tree, and how does it work in the context of classification?

Answer -

A Decision Tree is a supervised machine learning algorithm used for both classification and regression tasks, but it is more commonly applied in classification problems.

It is a tree-like structure where:

- Internal nodes represent a feature (attribute) on which data is split.

 - Branches represent a decision rule (condition).

 - Leaf nodes represent the final output (class label).

Think of it as a flowchart where you keep asking "Yes/No" or conditional questions until you arrive at a decision.

How it Works in Classification

1. Start at the Root Node
The process begins with the full dataset.

2. Select the Best Attribute to Split
The algorithm chooses the attribute that best separates the classes.

Criteria: Information Gain (Entropy reduction), Gini Index, or Chi-square.

3. Create Branches Based on Feature Values
Each branch represents a possible outcome of the attribute test.

4. Repeat Recursively
The process continues on each subset of data until:

 -All data points in a node belong to the same class, OR

 -There are no remaining features to split, OR

 -A stopping condition (like tree depth) is reached.

5. Assign Class Labels
The leaf node is labeled with the most frequent class in that subset.

Example

Suppose you want to predict if a person will buy insurance (Yes/No) based on: Age, Income, and Student status.

 - Root Node: Split on Age (best separator).

If Age < 30 → check Student.

If Age 30–50 → check Income.

If Age > 50 → likely Yes.

At the end of branches, each path leads to a classification decision.

Advantages

 - Easy to understand and visualize.

- Handles both numerical and categorical data.

 -No need for feature scaling.

Limitations

 - Can easily overfit (solution: pruning, ensemble methods like Random Forest).

 - Small changes in data can lead to different trees (unstable).

# Question 2: Explain the concepts of Gini Impurity and Entropy as impurity measures. How do they impact the splits in a Decision Tree?

ANSWER -

 Impurity Measures in Decision Trees -

When building a decision tree, we need a way to decide the best feature to split on at each step.
The idea:

 - A "pure" node = all samples belong to a single class (good split ✅).

 - An "impure" node = samples belong to multiple classes (bad split ❌).

To quantify this, we use Gini Impurity or Entropy.

 1. Gini Impurity

- Definition: Probability that a randomly chosen sample would be misclassified if it was randomly labeled according to the class distribution in that node.


        Gini=1−i=1∑C​(pi​)2

Where:   

        

𝐶
 = number of classes

𝑝
𝑖
	​= proportion of samples belonging to class
𝑖
i in the node

Key Intuition:

 - Lower Gini = purer node.

 - Minimum = 0 (perfect purity, all samples in one class).

 - Maximum = 0.5 (for binary classification, perfectly split 50–50).

Example:
If a node has 80% "Yes" and 20% "No":

             Gini=1−(0.82+0.22)=1−(0.64+0.04)=0.32


 2. Entropy (Information Gain)

 - Definition: Measures the uncertainty (disorder) in a node.

           Entropy=−i=1∑C​pi​⋅log2​(pi​)



Where:

𝑝
𝑖
	​
= proportion of class
𝑖


Key Intuition:

 - Entropy = 0 → pure node (all samples one class).

 - Higher entropy → more mixed classes.

 - Maximum = 1 (for binary classification with 50–50 split).

Example:
If a node has 80% "Yes" and 20% "No":

      Entropy=−(0.8log
2
	​

0.8+0.2log
2
	​

0.2)
=
−
(
0.8
⋅
0.32
+
0.2
⋅
2.32
)
≈
0.72
=−(0.8⋅0.32+0.2⋅2.32)≈0.72

Impact on Splits

 - Entropy → Information Gain

 - Decision Tree chooses the attribute that maximizes Information Gain:

         IG=Entropy(parent)−∑∣parent∣∣child∣​⋅Entropy(child)


 - Works well when class probabilities are very skewed.

    - Gini Impurity

 - CART (Classification and Regression Trees) typically use Gini.

 - Computationally faster than entropy.

 - Prefers splits that isolate the most frequent class quickly.

 Comparison

   | Measure           | Formula                 | Range             | When Used                                                     |
| ----------------- | ----------------------- | ----------------- | ------------------------------------------------------------- |
| **Gini Impurity** | (1 - \sum p_i^2)        | 0 to 0.5 (binary) | Faster, often default in CART                                 |
| **Entropy**       | (- \sum p_i \log_2 p_i) | 0 to 1            | Used in ID3, C4.5, gives more weight to less frequent classes |


# Question 3: What is the difference between Pre-Pruning and Post-Pruning in Decision Trees? Give one practical advantage of using each.

ANSWER -

🌳 Pruning in Decision Trees

Pruning = technique to reduce the size of a decision tree by removing unnecessary branches/nodes that do not provide significant predictive power.
It helps improve generalization on unseen data.

 1. Pre-Pruning (Early Stopping)

 - Definition: Stop growing the tree early, before it becomes overly complex.

 - How it works: Introduce constraints while building the tree, such as:

    - Maximum tree depth

     - Minimum number of samples required to split a node

     - Minimum information gain required for a split

    - Minimum samples per leaf

Advantage:
 Efficiency – Reduces computation time and prevents the tree from growing unnecessarily large, saving both time and memory.

 2. Post-Pruning (Prune After Full Growth)

 - Definition: Allow the tree to grow fully (possibly overfitting), then remove branches that do not improve performance.

 - How it works:

 - Grow the full tree.

 - (α-pruning) to cut back branches.

 - Usually evaluated on a validation set to check if pruning improves accuracy.

Advantage:
 Better Accuracy – Since the tree first learns all possible patterns, post-pruning can carefully remove only unhelpful splits, often leading to higher test accuracy compared to pre-pruning.

 Key Difference
   
   | Aspect           | Pre-Pruning                       | Post-Pruning                       |
| ---------------- | --------------------------------- | ---------------------------------- |
| **When Applied** | During tree construction          | After full tree construction       |
| **Control**      | Stops splits early                | Cuts unnecessary branches later    |
| **Risk**         | May underfit if stopped too early | Risk of overfitting before pruning |
| **Efficiency**   | Faster, saves resources           | Slower, needs full tree growth     |
| **Accuracy**     | Might miss patterns               | Usually better generalization      |


# Question 4: What is Information Gain in Decision Trees, and why is it important for choosing the best split?
ANSWER -

What is Information Gain?

- Definition:
Information Gain (IG) measures how much uncertainty (entropy) in the dataset is reduced after splitting on a particular feature.

In other words, it tells us how well a feature separates the data into distinct classes.

         IG(S,A)=Entropy(S)−v∈Values(A)∑​∣S∣∣Sv​∣​⋅Entropy(Sv​)

Where:

𝑆 = current dataset (parent node)

𝐴 = attribute we are splitting on

𝑆
𝑣 = subset of
𝑆
S for which attribute
𝐴
A takes value
𝑣
v

Entropy(S) = impurity before split

Weighted average entropy of children = impurity after split

 Why is it Important?

 - Decision Trees need a criterion to decide which feature to split on at each node.

 - Higher Information Gain = better split.

 - It ensures that:

    - The split maximizes purity of child nodes.

    - Reduces the "disorder" in the dataset.

    - Brings the tree closer to making a confident classification.

 Simple Example

Suppose we want to predict if students pass/fail based on study hours.

Before split (parent node):
Dataset: 10 students → 5 Pass, 5 Fail
     
        Entropy=−(0.5log2​0.5+0.5log2​0.5)=1
      

After split on “study hours > 5”:

 - Left child: 6 students → 5 Pass, 1 Fail → Entropy ≈ 0.65

 - Right child: 4 students → 0 Pass, 4 Fail → Entropy = 0

       
Weighted entropy =
6
10
(
0.65
)
+
4
10
(
0
)
=
0.39
10
6
	​

(0.65)+
10
4
	​

(0)=0.39

Information Gain:
      
       IG=1−0.39=0.61

 This means splitting on “study hours > 5” significantly reduces uncertainty, so it’s a good split.

  Key Points

Information Gain = reduction in entropy after a split.

Used to choose the most informative attribute at each node.

Helps tree grow in a way that maximizes class purity.

ID3 algorithm uses Information Gain as the splitting criterion.

# Question 5: What are some common real-world applications of Decision Trees, and what are their main advantages and limitations?

ANSWER -

Real-World Applications of Decision Trees

1. Finance & Banking

Loan approval / credit scoring → classifying whether a customer is likely to default.

Fraud detection → spotting unusual transactions.

2. Healthcare

Disease diagnosis → predicting whether a patient has a condition based on symptoms/test results.

Treatment recommendation → guiding doctors with patient profiles.

3. Marketing & Customer Analytics

Customer segmentation → classifying customers into buyer groups.

Churn prediction → predicting which customers may leave a service.

4. Retail & E-commerce

Product recommendation → classifying what a user is most likely to purchase.

Inventory management → predicting demand trends.

5. Human Resources

Candidate selection → classifying job applicants based on skills/experience.

Attrition prediction → which employees are likely to leave.

6. Manufacturing & Operations

Quality control → classifying items as defect / non-defect.

Supply chain decisions → risk analysis and logistics.

 Main Advantages of Decision Trees

1. Interpretability & Transparency

Easy to visualize and explain decisions (good for business users).

2. Handles Mixed Data Types

Works with both categorical and numerical features.

3. No Need for Feature Scaling

Unlike algorithms like SVM or KNN, no standardization required.

4. Non-Linear Relationships

Can capture complex decision boundaries.

5. Fast & Efficient

Works well on moderately sized datasets.

 Main Limitations of Decision Trees

1. Overfitting

Trees can become too deep and fit noise (solution: pruning, ensembles like Random Forest).

2. Instability

Small changes in data can produce very different trees.

3. Bias Toward Features with Many Levels

Features with many unique values (like IDs) may be chosen unfairly.

4. Limited Predictive Power Alone

Often outperformed by ensemble methods (Random Forest, Gradient Boosted Trees).

5. Not Always Optimal for Continuous Variables

Splits may not capture smooth relationships as well as regression models.



# Dataset Info:

●	Iris Dataset for classification tasks (sklearn.datasets.load_iris() or provided CSV).

●	Boston Housing Dataset for regression tasks (sklearn.datasets.load_boston() or provided CSV).
   
Question 6:  Write a Python program to:

●	Load the Iris Dataset

●	Train a Decision Tree Classifier using the Gini criterion

●	Print the model’s accuracy and feature importances (Include your Python code and output in the code box below.)

In [1]:
# Import required libraries
from sklearn.datasets import load_iris
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
import pandas as pd

# Load the Iris dataset
iris = load_iris()
X, y = iris.data, iris.target

# Split into training and testing sets (70% train, 30% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Train a Decision Tree Classifier using Gini criterion
clf = DecisionTreeClassifier(criterion="gini", random_state=42)
clf.fit(X_train, y_train)

# Make predictions
y_pred = clf.predict(X_test)

# Calculate accuracy
accuracy = accuracy_score(y_test, y_pred)

# Get feature importances
feature_importances = pd.Series(clf.feature_importances_, index=iris.feature_names)

accuracy, feature_importances


(1.0,
 sepal length (cm)    0.000000
 sepal width (cm)     0.019110
 petal length (cm)    0.893264
 petal width (cm)     0.087626
 dtype: float64)

Here’s the Python code and its output for training a Decision Tree Classifier on the Iris dataset using the Gini criterion

In [2]:
# Import required libraries
from sklearn.datasets import load_iris
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
import pandas as pd

# Load the Iris dataset
iris = load_iris()
X, y = iris.data, iris.target

# Split into training and testing sets (70% train, 30% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Train a Decision Tree Classifier using Gini criterion
clf = DecisionTreeClassifier(criterion="gini", random_state=42)
clf.fit(X_train, y_train)

# Make predictions
y_pred = clf.predict(X_test)

# Calculate accuracy
accuracy = accuracy_score(y_test, y_pred)

# Get feature importances
feature_importances = pd.Series(clf.feature_importances_, index=iris.feature_names)

print("Model Accuracy:", accuracy)
print("\nFeature Importances:\n", feature_importances)


Model Accuracy: 1.0

Feature Importances:
 sepal length (cm)    0.000000
sepal width (cm)     0.019110
petal length (cm)    0.893264
petal width (cm)     0.087626
dtype: float64


Interpretation:

 - The model achieved 100% accuracy on the test set (Iris is an easy dataset).

 - Petal length was the most important feature (~89.3%), followed by petal width (~8.7%).

 - Sepal length had no importance in the final splits.

# Question 7: Write a Python program to:

●	Load the Iris Dataset
●	Train a Decision Tree Classifier with max_depth=3 and compare its accuracy to a fully-grown tree.

(Include your Python code and output in the code box below.)

ANSWER -



In [4]:
# Import required libraries
from sklearn.datasets import load_iris
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

# Load the Iris dataset
iris = load_iris()
X, y = iris.data, iris.target

# Split into training and testing sets (70% train, 30% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Train a fully-grown Decision Tree Classifier
clf_full = DecisionTreeClassifier(random_state=42)
clf_full.fit(X_train, y_train)
y_pred_full = clf_full.predict(X_test)
accuracy_full = accuracy_score(y_test, y_pred_full)

# Train a Decision Tree Classifier with max_depth=3
clf_limited = DecisionTreeClassifier(max_depth=3, random_state=42)
clf_limited.fit(X_train, y_train)
y_pred_limited = clf_limited.predict(X_test)
accuracy_limited = accuracy_score(y_test, y_pred_limited)

print("Accuracy (Fully-grown tree):", accuracy_full)
print("Accuracy (Max depth=3):", accuracy_limited)


Accuracy (Fully-grown tree): 1.0
Accuracy (Max depth=3): 1.0


# Question 8: Write a Python program to:

●	Load the California Housing dataset from sklearn

●	Train a Decision Tree Regressor

●	Print the Mean Squared Error (MSE) and feature importances

(Include your Python code and output in the code box below.)

ANSWER -




In [5]:
# Import required libraries
from sklearn.datasets import fetch_california_housing
from sklearn.tree import DecisionTreeRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
import pandas as pd

# Load the California Housing dataset
housing = fetch_california_housing()
X, y = housing.data, housing.target

# Split into training and testing sets (70% train, 30% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Train a Decision Tree Regressor
regressor = DecisionTreeRegressor(random_state=42)
regressor.fit(X_train, y_train)

# Make predictions
y_pred = regressor.predict(X_test)

# Calculate Mean Squared Error (MSE)
mse = mean_squared_error(y_test, y_pred)

# Get feature importances
feature_importances = pd.Series(regressor.feature_importances_, index=housing.feature_names)

print("Mean Squared Error (MSE):", mse)
print("\nFeature Importances:\n", feature_importances)


Mean Squared Error (MSE): 0.5280096503174904

Feature Importances:
 MedInc        0.523456
HouseAge      0.052135
AveRooms      0.049418
AveBedrms     0.024974
Population    0.032206
AveOccup      0.139012
Latitude      0.089992
Longitude     0.088806
dtype: float64


# Question 9: Write a Python program to:

●	Load the Iris Dataset

●	Tune the Decision Tree’s max_depth and min_samples_split using GridSearchCV

●	Print the best parameters and the resulting model accuracy

(Include your Python code and output in the code box below.)

 ANSWER -





In [6]:
# Import required libraries
from sklearn.datasets import load_iris
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.metrics import accuracy_score

# Load the Iris dataset
iris = load_iris()
X, y = iris.data, iris.target

# Split into training and testing sets (70% train, 30% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Define parameter grid for tuning
param_grid = {
    'max_depth': [2, 3, 4, 5, None],
    'min_samples_split': [2, 3, 4, 5, 6]
}

# Initialize Decision Tree and GridSearchCV
clf = DecisionTreeClassifier(random_state=42)
grid_search = GridSearchCV(clf, param_grid, cv=5, scoring='accuracy')
grid_search.fit(X_train, y_train)

# Get best parameters
best_params = grid_search.best_params_

# Train best model
best_model = grid_search.best_estimator_
y_pred = best_model.predict(X_test)

# Calculate accuracy
accuracy = accuracy_score(y_test, y_pred)

print("Best Parameters:", best_params)
print("Model Accuracy:", accuracy)


Best Parameters: {'max_depth': 4, 'min_samples_split': 6}
Model Accuracy: 1.0


Interpretation:

The best model was found with max_depth=4 and min_samples_split=6.

It achieved 100% accuracy on the test set (since Iris is a simple dataset).

GridSearchCV helps ensure that the chosen parameters avoid overfitting and improve generalization.

# Question 10: Imagine you’re working as a data scientist for a healthcare company that wants to predict whether a patient has a certain disease. You have a large dataset with mixed data types and some missing values.
Explain the step-by-step process you would follow to:

●	Handle the missing values

●	Encode the categorical features

●	Train a Decision Tree model

●	Tune its hyperparameters

●	Evaluate its performance
And describe what business value this model could provide in the real-world



ANSWER -

Step-by-step process
1) Quick data check (always start here)

 - Quantify missingness per column and row, inspect patterns (heatmap / missingno), check distributions, class balance, and cardinality of categorical variables.

 - Decide whether missingness is MCAR / MAR / MNAR — that affects imputation choices and whether missingness itself is informative.

2) Handle missing values

Practical approach

 - If a column is >~50% missing, consider dropping it unless clinically important.

 - For numerical features:

    - Simple: median imputation (robust) inside a pipeline to avoid leakage.

    - Advanced: IterativeImputer (MICE) or KNN-imputer if relationships exist between variables.

    - Always add a missingness indicator (is_missing boolean column) when missingness might be informative.

 - For categorical features:

    - Impute with a new category like 'Missing' or the mode (but prefer explicit 'Missing' flag).

- Important: Do imputation inside the training pipeline / CV folds (never impute on full dataset before CV).

3) Encode categorical features

 - Binary: map to 0/1.

 - Nominal (low cardinality): OneHotEncoder(handle_unknown='ignore').

 - Ordinal (known order): OrdinalEncoder with explicit mapping.

 - High-cardinality (many categories): use target encoding or hashing. If using target encoding, use cross-validated target encoding (to avoid leakage) or implement within the CV pipeline.

 - NOTE for trees: tree models can often handle ordinal-encoded categories reasonably, but one-hot or target encoding are safer for nominal variables. Always avoid target leakage by computing encodings inside CV folds.

4) Train Decision Tree (safe reproducible workflow)

 - Split: use Stratified train/validation/test (preserve class ratios).

 - Build an sklearn Pipeline that includes:

    - ColumnTransformer for numeric/categorical preprocessing (imputer + encoder).

    - The DecisionTreeClassifier.

 - For class imbalance: either class_weight='balanced' in the classifier or perform resampling (SMOTE/undersampling) — but resample only inside CV folds.

 - Note: sklearn DecisionTreeClassifier does not accept NaNs, so ensure imputation step is present.

5) Hyperparameter tuning

 - Tune with RandomizedSearchCV or GridSearchCV using StratifiedKFold.

 - Important parameters to tune:

    - max_depth, min_samples_split, min_samples_leaf

    - max_features (None / 'sqrt' / 'log2' / fraction)

    - criterion ('gini' / 'entropy')

    - ccp_alpha (cost-complexity pruning)

    - class_weight (None / 'balanced')

 - Use an appropriate scoring: for disease detection you may prefer roc_auc or f1, or set custom scoring that prioritizes recall if missing the disease is costly.

 - Use nested CV if you must report unbiased generalization performance (outer CV to estimate performance, inner CV to tune).

6) Evaluate model performance (beyond one metric)

 - Standard metrics: confusion matrix, accuracy, precision, recall (sensitivity), F1.

 - For imbalanced or clinically important positive class: emphasize recall (sensitivity) and precision; report ROC-AUC and PR-AUC (PR-AUC is better when positives are rare).

 - Threshold tuning: choose probability threshold based on business tradeoffs (maximize recall at acceptable precision, or use cost matrix).

 - Calibration: check calibration (calibration_curve or CalibratedClassifierCV) if clinical decisions require well-calibrated probabilities.

 - Interpretability: show feature_importances_, visualize the tree, and produce SHAP values or partial dependence plots to explain predictions to clinicians.

 - Validation: evaluate on a hold-out test set and, if possible, an external dataset (temporal or other hospital).

 - Fairness & robustness: check performance across demographic subgroups.

7) Production considerations & monitoring

 - Explainability is critical in healthcare — keep human-in-the-loop (clinician review) and produce explanations for each prediction.

 - Operationalize with logging, model versioning, input validation, and monitoring for data drift and performance decay.

 - Clinical validation: clinical trials or retrospective validation are often required before real use.

 - Governance & privacy: adhere to regulations (HIPAA/GDPR-like rules), establish audit trails and informed consent where necessary.

 - Business value (what the model delivers)

 - Early detection / triage: identify high-risk patients earlier so clinicians can act faster → improved outcomes.

 - Resource allocation: prioritize tests and specialist referrals to patients most likely to benefit.

 - Cost savings: reduce unnecessary tests for low-risk patients; improve efficiency.

 - Decision support: assist clinicians with explainable risk scores, improving consistency and reducing human error.

 - Population health: identify at-risk cohorts for preventive interventions.

 - But: must balance false negatives (missed disease) vs false positives (unnecessary anxiety/cost). Good governance and clinician oversight are essential.



 Example sklearn pipeline

In [ ]:
import numpy as np
from sklearn.model_selection import train_test_split, StratifiedKFold, RandomizedSearchCV
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import OneHotEncoder, OrdinalEncoder
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import classification_report, roc_auc_score, precision_recall_curve

# X, y = ...  # load dataset
num_cols = ['age','bmi','systolic']         # example numeric cols
cat_cols = ['sex','smoker','ethnicity']     # example categorical cols

# Preprocessing
num_pipe = Pipeline([
    ('imputer', SimpleImputer(strategy='median')),
    # trees don't need scaling, so skipping scaler
])

cat_pipe = Pipeline([
    ('imputer', SimpleImputer(strategy='constant', fill_value='Missing')),
    ('onehot', OneHotEncoder(handle_unknown='ignore', sparse=False))
])

preproc = ColumnTransformer([
    ('num', num_pipe, num_cols),
    ('cat', cat_pipe, cat_cols)
])

# Full pipeline
pipe = Pipeline([
    ('preproc', preproc),
    ('clf', DecisionTreeClassifier(random_state=42))
])

# Hyperparam search (Randomized)
param_dist = {
    'clf__max_depth': [None, 3, 5, 7, 10],
    'clf__min_samples_split': [2, 5, 10, 20],
    'clf__min_samples_leaf': [1, 2, 4, 10],
    'clf__max_features': [None, 'sqrt', 'log2'],
    'clf__criterion': ['gini','entropy'],
    'clf__class_weight': [None, 'balanced'],
    'clf__ccp_alpha': [0.0, 0.001, 0.01]
}

cv = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)
search = RandomizedSearchCV(pipe, param_distributions=param_dist,
                            n_iter=50, cv=cv, scoring='roc_auc', n_jobs=-1, random_state=42)

search.fit(X_train, y_train)             # X_train/y_train created from stratified split
best = search.best_estimator_
y_pred = best.predict(X_test)
y_proba = best.predict_proba(X_test)[:,1]

print("Best params:", search.best_params_)
print(classification_report(y_test, y_pred))
print("ROC-AUC:", roc_auc_score(y_test, y_proba))
